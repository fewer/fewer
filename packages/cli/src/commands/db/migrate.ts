import { Command, flags } from '@oclif/command';
import commonFlags from '../../commonFlags';
import MigrationRunner from '../../MigrationRunner';
import { getMigrations, getDatabase } from '../../utils';

export default class Migrate extends Command {
  static description = 'Performs migrations.';

  static examples = [
    '$ fewer db:migrate',
    '$ fewer db:migrate --version 20190121190432',
    '$ fewer db:migrate --direction down --version 20190121190432',
    '$ fewer db:migrate --rollback',
    '$ fewer db:migrate --rollback --steps 3',
    '$ fewer db:migrate --redo --steps 3',
  ];

  static flags = {
    ...commonFlags,
    version: flags.string({
      char: 'v',
      description: 'Runs a specific migration that matches the version.',
    }),
    rollback: flags.boolean({
      description: 'Rolls back the previous migration.',
      exclusive: ['version'],
    }),
    redo: flags.boolean({
      description: 'Roll back a number of migrations, then perform them again.',
      exclusive: ['version'],
    }),
    steps: flags.integer({
      description: 'Sets the number of migrations to rollback, or redo.',
      exclusive: ['version'],
    }),
    direction: flags.string({
      description: 'Runs a specific migration in the provided direction.',
      options: ['up', 'down'],
      dependsOn: ['version'],
      exclusive: ['rollback', 'redo', 'steps'],
    }),
  };

  async run() {
    const { flags } = this.parse(Migrate);

    const dbFile = await getDatabase('Which database would you like to perform migrations on?');
    const migrations = await getMigrations(dbFile);

    const runner = new MigrationRunner(migrations);

    try {
      if (flags.redo) {
        await runner.redo(dbFile, flags.steps);
      } else if (flags.version) {
        if (flags.direction === 'down') {
          await runner.down(flags.version!);
        } else {
          await runner.up(flags.version!);
        }
      } else if (flags.rollback) {
        await runner.rollback(dbFile, flags.steps);
      } else {
        await runner.latest(dbFile);
      }
    } finally {
      await runner.cleanup();
    }
  }
}
