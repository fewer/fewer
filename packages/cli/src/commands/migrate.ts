import fs from 'fs';
import path from 'path';
import { promisify } from 'util';
import { Command, flags } from '@oclif/command';
import sortBy from 'lodash/sortBy';
import takeWhile from 'lodash/takeWhile';
import commonFlags from '../commonFlags';
import getConfig from '../getConfig';
import MigrationRunner from '../migrationRunner/Class';

const readdirAsync = promisify(fs.readdir);

export default class Migrate extends Command {
  static description = 'Performs migrations.';

  static examples = [
    '$ fewer migrate',
    '$ fewer migrate --version 20190121190432',
    '$ fewer migrate --direction down --version 20190121190432',
    '$ fewer migrate --rollback',
    '$ fewer migrate --rollback --steps 3',
    '$ fewer migrate --redo --steps 3',
  ];

  static flags = {
    ...commonFlags,
    version: flags.string({
      char: 'v',
      description: 'Runs a specific migration that matches the version.',
    }),
    rollback: flags.boolean({
      description: 'Rolls back the previous migration.',
      exclusive: ['version'],
    }),
    redo: flags.boolean({
      description: 'Roll back a number of migrations, then perform them again.',
      exclusive: ['version'],
    }),
    steps: flags.integer({
      description: 'Sets the number of migrations to rollback, or redo.',
      dependsOn: ['rollback', 'redo'],
    }),
    direction: flags.string({
      description: 'Runs a specific migration in the provided direction.',
      options: ['up', 'down'],
      dependsOn: ['version'],
      exclusive: ['rollback', 'redo', 'steps'],
    }),
  };

  // TODO: Move more of this into the migration runner.
  async run() {
    const { flags } = this.parse(Migrate);
    const config = await getConfig();

    // TODO: Store project root somewhere so that I don't need to do cwd garbage:
    const migrationFiles = await readdirAsync(
      path.resolve(process.cwd(), config.migrations),
    );

    const migrations = sortBy(
      migrationFiles.filter(filename => !filename.startsWith('.')),
    ).map(filename => path.join(process.cwd(), config.migrations, filename));

    const runner = new MigrationRunner(migrations);

    if (flags.redo) {
      runner.redo(flags.steps);
    } else if (flags.version) {
      if (flags.direction === 'down') {
        runner.down(flags.version!);
      } else {
        runner.up(flags.version!);
      }
    } else if (flags.rollback) {
      runner.rollback(flags.steps);
    } else {
      runner.latest();
    }

    await runner.go();
  }
}
