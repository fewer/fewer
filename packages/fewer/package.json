{
  "name": "fewer",
  "version": "0.0.6",
  "description": "A minimal ORM for Node.js.",
  "keywords": [
    "orm",
    "sql",
    "fewer"
  ],
  "author": "Jordan Gensler <jgensler@netflix.com>",
  "homepage": "https://github.com/fewer/fewer",
  "license": "MIT",
  "main": "lib/index.js",
  "directories": {
    "lib": "lib",
    "test": "__tests__"
  },
  "files": [
    "lib"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/fewer/fewer.git"
  },
  "scripts": {
    "test": "echo \"Error: run tests from root\" && exit 1"
  },
  "bugs": {
    "url": "https://github.com/fewer/fewer/issues"
  },
  "dependencies": {
    "@fewer/sq": "^0.0.4"
  },
  "devDependencies": {
    "@fewer/typeval": "^0.0.4"
  }
}
