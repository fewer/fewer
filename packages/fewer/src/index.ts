import Schema from './Schema';
import { Repository, createRepository } from './Repository';

export { Schema, Repository, createRepository };
