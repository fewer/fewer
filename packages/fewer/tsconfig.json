{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "lib",
    "composite": true,
    "rootDir": "src"
  },
  "include": ["./src"],
  "references": [
    { "path": "../sq" },
  ]
}
