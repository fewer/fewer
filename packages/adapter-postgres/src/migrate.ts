import { Migration, ColumnType, Operations } from 'fewer';
import squel from './squel';
import { CharacterOptions, NumericOptions } from './columnTypes';

function getTypeName(type: ColumnType): string {
  // If there is no config, then we only have the name:
  if (!type.config) return type.name;

  switch (type.name) {
    case 'char':
    case 'varchar': {
      const config: CharacterOptions = type.config;
      if (config.length) {
        return `${type.name}(${config.length})`;
      }
      return type.name;
    }
    case 'numeric':
    case 'decimal': {
      const config: NumericOptions = type.config;
      if (config.scale) {
        return `${type.name}(${config.precision}, ${config.scale})`;
      }
      if (config.precision) {
        return `${type.name}(${config.precision})`;
      }
      return type.name;
    }
    case 'array': {
      return `${getTypeName(type.config.subtype)}[]`;
    }
    default:
      return type.name;
  }
}

function createTable(operation: Operations.CreateTable) {
  const create = squel.create(operation.options);
  create.table(operation.name);
  for (const [columnName, columnType] of Object.entries(operation.columns)) {
    const typeName = getTypeName(columnType);
    create.field(columnName, typeName, columnType.config);
  }
  return create.toString();
}

function dropTable(operation: Operations.DropTable) {
  return squel.dropTable(operation.name).toString();
}

export default function migrate(
  migration: Migration,
) {
  const sqls: (null | string)[] = migration.operations.map(operation => {
    switch (operation.type) {
      case 'createTable':
        return createTable(operation);
      case 'dropTable':
        return dropTable(operation);
      default:
        return null;
    }
  });

  return sqls.filter(Boolean).join('; ');
}
