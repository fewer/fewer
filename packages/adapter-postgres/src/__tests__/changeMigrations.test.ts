import { createMigration, createDatabase } from 'fewer';
import { Adapter } from '../';
import migrate from '../migrate';

const database = createDatabase({
  adapter: new Adapter({}),
});

const VERSION = 1;

describe('migration', () => {
  describe('createTable Down', () => {
    it('generates drop table sql', () => {
      const migration = createMigration(VERSION, database, {
        change: (m, t) =>
          m.createTable('users', null, {
            id: t.bigserial({ primaryKey: true }),
            firstName: t.string(),
            lastName: t.string(),
            email: t.string({ nonNull: true }),
          }),
      });

      migration.prepare('up');

      const sql = migrate(migration);
      expect(sql).toMatchSnapshot();
    });
  });

  describe('createTable Up', () => {
    it('generates create table sql', () => {
      const migration = createMigration(VERSION, database, {
        change: (m, t) =>
          m.createTable('users', null, {
            id: t.bigserial({ primaryKey: true }),
            firstName: t.string(),
            lastName: t.string(),
            email: t.string({ nonNull: true }),
          }),
      });

      migration.prepare('up');

      const sql = migrate(migration);
      expect(sql).toMatchSnapshot();
    });

    it('supports unique constraint', () => {
      const migration = createMigration(VERSION, database, {
        change: (m, t) =>
          m.createTable('users', null, {
            id: t.bigserial({ primaryKey: true }),
            firstName: t.string(),
            lastName: t.string(),
            email: t.string({ unique: true, nonNull: true }),
          }),
      });

      migration.prepare('up');

      const sql = migrate(migration);
      expect(sql).toMatchSnapshot();
    });

    it('supports compound primary key', () => {
      const migration = createMigration(VERSION, database, {
        change: (m, t) =>
          m.createTable(
            'users',
            { primaryKey: ['id', 'email'] },
            {
              id: t.bigserial(),
              firstName: t.string(),
              lastName: t.string(),
              email: t.string({ nonNull: true }),
            },
          ),
      });

      migration.prepare('up');

      const sql = migrate(migration);
      expect(sql).toMatchSnapshot();
    });
  });
});
