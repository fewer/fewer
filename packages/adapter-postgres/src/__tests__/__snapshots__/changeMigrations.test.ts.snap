// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`migration createTable Down generates drop table sql 1`] = `"CREATE TABLE users (id BIGSERIAL PRIMARY KEY, firstName VARCHAR(255), lastName VARCHAR(255), email VARCHAR(255) NOT NULL)"`;

exports[`migration createTable Up generates create table sql 1`] = `"CREATE TABLE users (id BIGSERIAL PRIMARY KEY, firstName VARCHAR(255), lastName VARCHAR(255), email VARCHAR(255) NOT NULL)"`;

exports[`migration createTable Up supports compound primary key 1`] = `"CREATE TABLE users (id BIGSERIAL, firstName VARCHAR(255), lastName VARCHAR(255), email VARCHAR(255) NOT NULL)"`;

exports[`migration createTable Up supports unique constraint 1`] = `"CREATE TABLE users (id BIGSERIAL PRIMARY KEY, firstName VARCHAR(255), lastName VARCHAR(255), email VARCHAR(255) NOT NULL UNIQUE)"`;
